resolver/addressable_resolver.go

```diff
@@ -56,6 +58,28 @@ func NewURIResolver(ctx context.Context, callback func(types.NamespacedName)) *U
 		},
 	}
 
+	ret.listerFactory = func(gvr schema.GroupVersionResource) (cache.GenericLister, error) {
+		_, l, err := ret.informerFactory.Get(ctx, gvr)
+		return l, err
+	}
+
+	return ret
+}
+
+// NewURIResolver constructs a new URIResolver with context and a callback
+// for a given listableType (Listable) passed to the URIResolver's tracker.
+// extra when controller context is used
+func NewURIResolverWithControllerContext(ctx context.Context, dmInformerFactory pkgapisduck.InformerFactory,
```

Comment:
I'd really like to see some test coverage for this, especially in order to be able to see how the API would look for consumers (a WIP version of the serving PR might be good for this too). It might also be worth adding a test for the new behaviour here (that cancelling the context passed to `URIFromObjectReference` etc doesn't cause problems for future calls to the resolver), though looks like I didn't do that either in https://github.com/knative/serving/pull/10673 🤭.